x/programs/rust/wasmlanche_sdk/src/simulator.rs

```diff
@@ -0,0 +1,326 @@
+//! A client and types for the VM simulator. This feature allows for Rust
+//! developers to construct tests for their programs completely in Rust.
+//! Alternatively the `Plan` can be written in YAML/JSON and passed to the
+//! Simulator binary directly.
+
+use std::{
+    error::Error,
+    io::Write,
+    process::{Command, Stdio},
+};
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+#[serde(rename_all = "lowercase")]
+pub enum Endpoint {
+    /// Perform an operation against the key api.
+    Key,
+    /// Make a read-only call to a program function and return the result.
+    ReadOnly,
+    /// Create a transaction on-chain from a possible state changing program
+    /// function call. A program's function can internally optionally call other
+    /// functions including program to program.
+    Execute,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct Step<'a> {
+    /// A description of the step.
+    description: &'a str,
+    /// The API endpoint to call.
+    endpoint: Endpoint,
+    /// The method to call on the endpoint.
+    method: &'a str,
+    /// The parameters to pass to the method.
+    params: Vec<Param<'a>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    require: Option<Require>,
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "lowercase")]
+pub enum ParamType {
+    U64,
+    String,
+    Id,
+    #[serde(untagged)]
+    Key(Key),
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+#[serde(rename_all = "lowercase")]
+pub enum Key {
+    Ed25519,
+    Secp256r1,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct Param<'a> {
+    /// The optional name of the parameter. This is only used for readability.
+    name: &'a str,
+    #[serde(rename = "type")]
+    /// The type of the parameter.
+    param_type: ParamType,
```

Comment:
the parameter type must be specified so json knows how to parse?

Reply:
yeah I want to use `type` but its reserved in rust and r#type is even more annoying to type :)